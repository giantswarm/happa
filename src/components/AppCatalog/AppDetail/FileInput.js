import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React from 'react';

const Wrapper = styled.div`
  margin-bottom: 15px;
  :last-child: {
    margin-bottom: 0;
  }
`;

const Text = styled.div`
  font-size: 14px;
  color: ${(props) => props.theme.colors.white2};
`;

const InputWrapper = styled.div`
  align-items: center;
  display: flex;
`;

const Input = styled.input`
  background-color: ${(props) => props.theme.colors.shade5};
  border: 1px solid ${(props) => props.theme.colors.shade6};
  border-radius: ${(props) => props.theme.border_radius};
  font-size: 14px;
  line-height: normal;
  padding: 8px 10px;
  width: 100%;
`;

const Icon = styled.i`
  color: ${(props) => props.theme.colors.white3};
  font-size: 24px;
  margin-right: 5px;
`;

const ValidationError = styled.span`
  font-size: 12px;
  color: ${(props) => props.theme.colors.yellow1};
`;

const Hint = styled.span`
  font-size: 12px;
`;

const FileInput = (props) => {
  const onChange = (e) => {
    if (props.onChange) {
      props.onChange(e.target.files);
    }
  };

  return (
    <Wrapper>
      <Text>
        {props.label ? (
          <label htmlFor={props.label}>{props.label}</label>
        ) : undefined}
        {props.description ? <p>{props.description}</p> : undefined}
      </Text>
      <InputWrapper>
        {props.icon ? <Icon className={`fa fa-${props.icon}`} /> : undefined}
        <Input id={props.label} onChange={onChange} type='file' />
      </InputWrapper>
      {props.validationError ? (
        <ValidationError>
          <i className='fa fa-warning' /> {props.validationError}
        </ValidationError>
      ) : (
        <Hint>{props.hint}</Hint>
      )}
    </Wrapper>
  );
};

FileInput.propTypes = {
  description: PropTypes.string,
  hint: PropTypes.any,
  icon: PropTypes.string,
  label: PropTypes.string,
  onChange: PropTypes.func,
  validationError: PropTypes.string,
  value: PropTypes.object,
};

export default FileInput;
