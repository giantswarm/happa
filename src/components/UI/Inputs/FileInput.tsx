import PropTypes from 'prop-types';
import React, { ChangeEvent, ElementRef, FC } from 'react';

import Input, { IInput, InputElement } from './Input';

const FileInput: FC<IInput<FileList>> = (props) => {
  const onChange = (e: ChangeEvent<ElementRef<'input'>>) => {
    // eslint-disable-next-line no-unused-expressions
    props.onChange?.(e.target.files);
  };

  return (
    <Input {...props}>
      <InputElement id={props.label} onChange={onChange} type='file' />
    </Input>
  );
};

/*
  PropTypes wierdness, redefining `value` which is already `any`
  again as `any`.
 */
FileInput.propTypes = { ...Input.propTypes, value: PropTypes.any };

export default FileInput;
