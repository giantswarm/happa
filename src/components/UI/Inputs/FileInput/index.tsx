// TODO(axbarsan): Replace with `FileInput` once it's released.
import { FormField, FormFieldProps, TextInput as Input } from 'grommet';
import PropTypes from 'prop-types';
import React, { useEffect, useRef } from 'react';

interface IFileInputProps
  extends Omit<React.ComponentPropsWithRef<typeof Input>, 'type' | 'value'> {
  /**
   * The description text displayed above the input.
   */
  label?: FormFieldProps['label'];
  /**
   * Props to be passed to the text input wrapper.
   */
  contentProps?: FormFieldProps['contentProps'];
  /**
   * Props to be passed to the form field.
   */
  formFieldProps?: FormFieldProps & { className?: string };
  /**
   * Whether the input is required or not.
   */
  required?: FormFieldProps['required'];
  /**
   * An error to be displayed below the input.
   */
  error?: FormFieldProps['error'];
  /**
   * A description to be displayed below the input.
   */
  info?: FormFieldProps['info'];
  /**
   * A help text to be displayed between the label
   * and the input.
   */
  help?: FormFieldProps['help'];
  /**
   * The margin between the form field and other elements.
   */
  margin?: FormFieldProps['margin'];
  /**
   * Whether the input should have some extra padding or not.
   */
  pad?: FormFieldProps['pad'];
  value?: FileList | null;
}

const FileInput = React.forwardRef<HTMLInputElement, IFileInputProps>(
  (
    {
      id,
      label,
      contentProps,
      formFieldProps,
      disabled,
      required,
      error,
      info,
      help,
      name,
      margin,
      pad,
      children,
      value,
      ...props
    },
    ref
  ) => {
    const inputRef = useRef<HTMLInputElement | null>(null);

    useEffect(() => {
      if (inputRef.current) {
        inputRef.current.files = value!;
      }
    }, [value]);

    return (
      <FormField
        htmlFor={id}
        label={label}
        contentProps={contentProps}
        disabled={disabled}
        required={required}
        name={name}
        error={error}
        info={info}
        help={help}
        margin={margin}
        pad={pad}
        {...formFieldProps}
      >
        <Input
          type='file'
          id={id}
          disabled={disabled}
          required={required}
          name={name}
          {...props}
          ref={(element) => {
            inputRef.current = element;

            // TODO(axbarsan): Extract into utility.
            if (typeof ref === 'function') {
              ref(element);
            } else if (ref) {
              ref.current = element;
            }
          }}
        />
        {children}
      </FormField>
    );
  }
);

FileInput.propTypes = {
  id: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  contentProps: PropTypes.object,
  formFieldProps: PropTypes.object,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  info: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  help: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  name: PropTypes.string,
  margin: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  pad: PropTypes.bool,
  children: PropTypes.node,
  value: PropTypes.object as PropTypes.Requireable<FileList>,
};

FileInput.defaultProps = {
  size: 'medium',
  value: null,
};

export default FileInput;
