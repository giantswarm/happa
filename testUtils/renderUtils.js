import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
