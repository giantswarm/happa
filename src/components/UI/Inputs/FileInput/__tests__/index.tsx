import { renderWithTheme } from 'test/renderUtils';

import FileInput from '..';

describe('FileInput', () => {
  it('renders a simple input', () => {
    const { container } = renderWithTheme(FileInput, {});
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a label', () => {
    const { container } = renderWithTheme(FileInput, {
      id: 'input',
      label: 'Some input',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a validation error', () => {
    const { container } = renderWithTheme(FileInput, {
      error: 'You failed at this',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a help message', () => {
    const { container } = renderWithTheme(FileInput, {
      help: 'This is a very cool input',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with an info message', () => {
    const { container } = renderWithTheme(FileInput, {
      info: 'You can set a cool value',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a disabled input', () => {
    const { container } = renderWithTheme(FileInput, {
      disabled: true,
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
