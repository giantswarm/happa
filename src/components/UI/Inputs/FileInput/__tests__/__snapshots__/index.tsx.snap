// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput renders a disabled input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 iUQqn"
>
  <div
    class="sc-iCoGMd hFJToB"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hcMLLM FormField__FormFieldBox-m9hood-0 ofKvz"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 cwwoTn FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 gXHnix"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 zyHQr"
            disabled=""
            type="file"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FileInput renders a simple input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 iUQqn"
>
  <div
    class="sc-iCoGMd hFJToB"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hcMLLM FormField__FormFieldBox-m9hood-0 ofKvz"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 SMFjn FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 gXHnix"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 gHAWHi"
            type="file"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FileInput renders an input with a help message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 iUQqn"
>
  <div
    class="sc-iCoGMd hFJToB"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hcMLLM FormField__FormFieldBox-m9hood-0 ofKvz"
    >
      <span
        class="StyledText-sc-1sadyjn-0 bmnpwp"
      >
        This is a very cool input
      </span>
      <div
        class="StyledBox-sc-13pk1d4-0 SMFjn FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 gXHnix"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 gHAWHi"
            type="file"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FileInput renders an input with a label 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 iUQqn"
>
  <div
    class="sc-iCoGMd hFJToB"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hcMLLM FormField__FormFieldBox-m9hood-0 ofKvz"
    >
      <label
        class="StyledText-sc-1sadyjn-0 ZJQGZ"
        for="input"
      >
        Some input
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 SMFjn FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 gXHnix"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 gHAWHi"
            id="input"
            type="file"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FileInput renders an input with a validation error 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 iUQqn"
>
  <div
    class="sc-iCoGMd hFJToB"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hcMLLM FormField__FormFieldBox-m9hood-0 ofKvz"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 HhzTD FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 gXHnix"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 gHAWHi"
            type="file"
            value=""
          />
        </div>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 fnpEr"
      >
        You failed at this
      </span>
    </div>
  </div>
</div>
`;

exports[`FileInput renders an input with an info message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 iUQqn"
>
  <div
    class="sc-iCoGMd hFJToB"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 hcMLLM FormField__FormFieldBox-m9hood-0 ofKvz"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 SMFjn FormField__FormFieldContentBox-m9hood-1"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 gXHnix"
        >
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 gHAWHi"
            type="file"
            value=""
          />
        </div>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 bTUvUa"
      >
        You can set a cool value
      </span>
    </div>
  </div>
</div>
`;
